### PR TITLE
Fix camera view/flight issues when colliding with terrain

### LINKED ISSUE #4075: Camera flight issues
The below Sandcastle example is a simple save/restore camera view that exposes two issues in Cesium.

1. Flights that start far away from their destination and end up on the ground will cause the camera to collide with terrain while it is in flight. This causes the camera to end up wherever the terrain was loaded when it collided and not the end destination.

2. I think the minimum distance is being taken into account differently during camera flights than it does during camera motion. This means if you move the camera close to the ground with a horizon view, save the settings, and then do a flight; you end up in a difference location.

```javascript
var viewer = new Cesium.Viewer('cesiumContainer');
var cesiumTerrainProviderMeshes = new Cesium.CesiumTerrainProvider({
    url : 'https://assets.agi.com/stk-terrain/world',
    requestWaterMask : true,
    requestVertexNormals : true
});
viewer.terrainProvider = cesiumTerrainProviderMeshes;

var view;
var camera = viewer.scene.camera;

Sandcastle.addToolbarButton('Store View', function(){
    view = {
        destination : Cesium.Cartesian3.clone(camera.positionWC),
        orientation : {
            heading: camera.heading,
            pitch: camera.pitch,
            roll: camera.roll
        }
    };
});

Sandcastle.addToolbarButton('Load View', function(){
    camera.flyTo(view);
});
```

Using the above code, fly to San Francisco and create a horizon view right up against the ground.  Storing and loading the view won't match.

Stored view
image

Result after flight
image

### PR DESCRIPTION
**Pull request by bagnell**
NOTE: This is not ready for code review. I'm opening it to get feedback for if this is actually what we want to do. If we decide to go with it, I'll clean up the code and add tests.

Disable camera input before a flight and do not re-enable until the terrain has finished loading for the area.

Fixes #4075.

### CODE REVIEW

**Comment by mramato** [comment]
This seems to be completely broken.  When I use the sample code from #4075, the camera rotates, but does not move.

**Comment by bagnell** [comment]
I forgot to mention that there is a problem with the code sample. You need to clone the camera position, not store the reference or the position will stay the same.

**Comment by mramato** [comment]
OK, thanks, I fixed the example so no one makes the same mistake again.

**Comment by mramato** [comment]
This is definitely better in that it recreates the view properly, the only (major) downside is that since it waits for _all_ tiles to load, the camera can appear "stuck" for several seconds (5 in my case) while tiles are rendering.  Perhaps this is unavoidable given our current loading architecture?  Are we actually waiting for all tiles, or just all tile sunder the camera position?

@pjcozzi since you reported this problem, I would recommend you test this out.  The problem is definitely gone, but flying from the home view to a ground horizon view (which is the worse case scenario) definitely has a long delay.

**Comment by pjcozzi** [comment]
@bagnell there are JSHint warnings: https://travis-ci.org/AnalyticalGraphicsInc/cesium/builds/144003813

**Comment by pjcozzi** [comment]
> @pjcozzi since you reported this problem, I would recommend you test this out. The problem is definitely gone, but flying from the home view to a ground horizon view (which is the worse case scenario) definitely has a long delay.

Sounds bad, even with Network throttling, I did not notice this.  After I zoomed to the horizon view, I was able to pan even while the tiles (maybe just imagery?!?!?) load in.

**Comment by mramato** [comment]
> Sounds bad, even with Network throttling, I did not notice this. After I zoomed to the horizon view, I was able to pan even while the tiles (maybe just imagery?!?!?) load in.

Did you have terrain on?  From my laptop right now I have major lag.  I can demo for you tomorrow.

**Comment by pjcozzi** [comment]
Yes, with terrain.

@bagnell do you have a similar lag?

**Comment by bagnell** [comment]
It takes about 3-4 seconds before I can move the camera again when flying from the home view.

>Are we actually waiting for all tiles, or just all tile sunder the camera position?

Its waiting for all tiles. The other way to do it would be to repeatedly try to pick the terrain under the camera. As soon as some terrain is picked, we can unlock the camera. The down side is that the camera could really be positioned well below the terrain so it may never get picked.

**Comment by bagnell** [comment]
I changed it to only disable collision detection until the terrain is loaded not input like we talked about offline.

@pjcozzi @mramato Let me know what you think. I think its worse.

**Comment by pjcozzi** [comment]
Up to @mramato if we want to take this; otherwise, we can wait a bit for a real fix that prefetches terrain at the flight's target.

**Comment by mramato** [comment]
I think the latest behavior is a good compromise until we refactor terrain and imagery loading.

**Comment by pjcozzi** [comment]
I agree, this will get it done!

@bagnell 
* There are JSHint warnings: https://travis-ci.org/AnalyticalGraphicsInc/cesium/builds/144830985
* Update CHANGES.md
* Add tests if reasonable

**Comment by bagnell** [comment]
@pjcozzi @mramato This is ready for another review.

File: Source/Scene/Camera.js [hunk 1]
```diff
@@ -338,6 +339,77 @@ define([
         }
     };
 
+    var scratchAdjustHeightTransform = new Matrix4();
+    var scratchAdjustHeightCartographic = new Cartographic();
+
+    Camera.prototype._adjustHeightForTerrain = function() {
+        var scene = this._scene;
+
+        // Should these be moved to the camera?
```
Comment (pjcozzi) [inline Source/Scene/Camera.js]: Remove this comment and submit an issue for this if you think it is worth it.

**Comment by pjcozzi** [comment]
@mramato can you please test and merge when happy?

**Comment by mramato** [comment]
Looks good to me, once @pjcozzi's one comment is addressed, I'll merge.

**Comment by bagnell** [comment]
@mramato This is ready.

**Comment by mramato** [comment]
Thanks!

**Pull request merged**